bitacross-worker/bitacross/core/bc-musig2-ceremony/src/lib.rs

```diff
@@ -28,74 +28,60 @@ use std::{format, string::String, sync::Arc};
 #[cfg(all(feature = "std", feature = "sgx"))]
 compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
 
-use k256::SecretKey;
-use musig2::{
-	secp::Point, verify_single, BinaryEncoding, CompactSignature, KeyAggContext, LiftedSignature,
-	SecNonceSpices,
-};
-use std::{vec, vec::Vec};
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+
+#[cfg(feature = "sgx")]
+use std::sync::SgxRwLock as RwLock;
 
 use crate::CeremonyEvent::CeremonyEnded;
 use codec::{Decode, Encode};
 use itp_sgx_crypto::{key_repository::AccessKey, schnorr::Pair as SchnorrPair};
+use k256::SecretKey;
 pub use k256::{elliptic_curve::sec1::FromEncodedPoint, PublicKey};
 use litentry_primitives::RequestAesKey;
-use log::{debug, error, info, trace};
-use musig2::secp::Scalar;
+use log::*;
+use musig2::{
+	secp::{Point, Scalar},
+	verify_single, BinaryEncoding, CompactSignature, KeyAggContext, LiftedSignature,
+	SecNonceSpices,
+};
 pub use musig2::{PartialSignature, PubNonce};
-use std::collections::HashMap;
+use std::{
+	collections::HashMap,
+	time::{SystemTime, UNIX_EPOCH},
+	vec,
+	vec::Vec,
+};
 
 pub type CeremonyId = SignBitcoinPayload;
 pub type SignaturePayload = Vec<u8>;
 pub type Signers = Vec<SignerId>;
-pub type CeremonyRegistry<AK> = HashMap<CeremonyId, MuSig2Ceremony<AK>>;
-pub type CeremonyCommandsRegistry = HashMap<CeremonyId, Vec<PendingCeremonyCommand>>;
+pub type CeremonyRegistry<AK> = HashMap<CeremonyId, (Arc<RwLock<MuSig2Ceremony<AK>>>, u64)>;
+pub type CeremonyCommandTmp = HashMap<CeremonyId, (Arc<RwLock<Vec<CeremonyCommand>>>, u64)>;
 // enclave public key is used as signer identifier
 pub type SignerId = [u8; 32];
 pub type SignersWithKeys = Vec<(SignerId, PublicKey)>;
 
-pub struct PendingCeremonyCommand {
-	pub ticks_left: u32,
-	pub command: CeremonyCommand,
-}
-
-impl PendingCeremonyCommand {
-	pub fn tick(&mut self) {
-		self.ticks_left -= 1;
-	}
-}
-
-#[derive(Encode, Debug)]
-pub enum SignBitcoinError {
-	InvalidSigner,
-	CeremonyError,
-}
-
-#[derive(Debug, Eq, PartialEq)]
+#[derive(Debug, Eq, PartialEq, Encode)]
 pub enum CeremonyError {
-	NonceReceivingError(NonceReceivingErrorReason),
-	PartialSignatureReceivingError(PartialSignatureReceivingErrorReason),
+	CeremonyInitError(CeremonyErrorReason),
+	NonceReceivingError(CeremonyErrorReason),
+	PartialSignatureReceivingError(CeremonyErrorReason),
 }
 
-#[derive(Debug, Eq, PartialEq)]
-pub enum NonceReceivingErrorReason {
+#[derive(Debug, Eq, PartialEq, Encode)]
+pub enum CeremonyErrorReason {
 	SignerNotFound,
 	ContributionError,
 	IncorrectRound,
-	FirstRoundFinalizationError,
-}
-
-#[derive(Debug, Eq, PartialEq)]
-pub enum PartialSignatureReceivingErrorReason {
-	SignerNotFound,
-	ContributionError,
-	IncorrectRound,
-	SecondRoundFinalizationError,
+	RoundFinalizationError,
 }
 
 // events come from outside and are consumed by ceremony in tick fn
-#[derive(Debug)]
+#[derive(Debug, PartialEq, Eq, Clone)]
 pub enum CeremonyCommand {
+	Init,
```

Comment:
I'm wondering if we can avoid this command. Currently it just produces event - i think event should be created alongside ceremony.

Reply:
I added this command to unify processing flow, now every request will generate a command, and all events are generated by these commands.
'DirectCall::SignBitcoin' -> "CeremonyCommand::Init" -> "CeremonyEvent::FirstRoundStarted"
'DirectCall::NonceShare' -> "CeremonyCommand::SaveNonce" -> "CeremonyEvent::SecondRoundStarted"
'DirectCall::PartialSignatureShare' -> "CeremonyCommand::SavePartialSignature" -> "CeremonyEvent::CeremonyEnded"

Reply:
You made it work but it doesn't fit well after your code changes. 

I think you should made further changes:

1. remove command processing from ceremony
2. enrich `InitCeremony` with all fields required for ceremony creation
3. make `sign_bitcoin` direct call handler dumb, so it will only create and return `InitCeremony` command
4. create cereomony during `InitCeremony` command processing in `bc-task-processor` 

This will lead to more consistent design.

Reply:
> 1. remove command processing from ceremony

Could you explain more about this?

Reply:
Done